sklearn/ensemble/_weight_boosting.py

```diff
@@ -138,6 +148,28 @@ def fit(self, X, y, sample_weight=None):
         # Check parameters
         self._validate_estimator()
 
+        if is_regressor(self):
+            routed_params = process_routing(
+                obj=self, method="fit", other_params=fit_params
+            )
+        else:
+            # this will fail in AdaBoostClassifier if base_estimator has not
+            # requested `sample_weight`. we accept this as a requirement since
+            # it's a requirement for AdaBoost for the weak estimator to always
+            # accept `sample_weight`.
+            routed_params = process_routing(
+                obj=self,
+                method="fit",
+                other_params=fit_params,
+                sample_weight=sample_weight,
+            )
+
+        # we then remove sample_weight from the routed parameters since we will
+        # pass it explicitly to the fit method of the base_estimator. It was
+        # included above only to have `process_routing` to do the validation.
+        if "sample_weight" in routed_params.base_estimator.fit:
+            del routed_params.base_estimator.fit["sample_weight"]
```

Comment:
I'm not sure which one's best here?

- we pass sample_weight to routing as is implemented here, and it would raise if the user hasn't requested sample weight for their base_estimator, even if they haven't provided sample_weight. While working with @glemaitre we thought this makes sense since AdaBoostClassifier always passes sample_weight to the base_estimator, which means the user should always request it explicitly.
- we could only pass sample_weight to routing if it's explicitly provided by the user. This would make users' code not change in most cases since they mostly pass no explicit sample_weight, and they pass a simple consumer estimator as the base_estimator. Their code would still break if they pass a router as a base_estimator and the sub-estimators of that router are not requesting sample_weight.

Which one do you think makes more sense?

cc @thomasjpfan @jnothman

Reply:
I would prefer always requiring the request to be consistent.

If we want to be extra lenient, I say the `AdaBoost` will set the request of the `base_estimator` if the `sample_weight` request is not set yet. Assuming `base_estimator` is not requesting `sample_weight`, we can be consistent behavior to `main`:

1. If `base_estimator` is a classifier, `AdaBoostClassifer` sets the metadata request to True.
2. If `base_estimator` is a regressor, `AdaBoostClasifer` sets the meatadata request to False.

Reply:
We can't do that since _routers_ don't expose a method to set requests. So that would only work if the user passes a simple _consumer_ estimator.